### PR TITLE
feat: sort telemetry exports by timestamp

### PR DESCRIPTION
**Pull request by EhabY**
## Summary

Telemetry exports previously sorted same-day files by filename session slug, so events from different sessions on the same date could appear out of chronological order.

- Sort telemetry exports by event timestamp across same-day sessions instead of filename session slug.
- Keep export streaming by merging one event cursor per session.
- Preserve append order when a single session emits backward timestamps.
- Keep filename format, schema, retention, and support bundle behavior unchanged.

## Commits

- `feat: sort telemetry exports by timestamp`: initial behavior change.
- `fix: address telemetry export review feedback`: review fixes and test coverage cleanup.

## Validation

- `pnpm test:extension ./test/unit/telemetry/export/files.test.ts ./test/unit/telemetry/export/pipeline.test.ts`
- `pnpm typecheck`
- `pnpm format:check`
- `pnpm lint`

Generated by Coder Agents.

### CODE REVIEW

File: src/telemetry/export/files.ts [hunk 1]
```diff
@@ -50,17 +62,44 @@ export async function listTelemetryFilesForRange(
...
-		const name = path.basename(filePath);
-		const stream = createReadStream(filePath, { encoding: "utf8" });
+	const frontier: EventCursor[] = [];
+	for (const files of sessionFileGroups(filePaths)) {
+		const iterator = streamTelemetryEventEntries(files, range)[
+			Symbol.asyncIterator
+		]();
+		const next = await iterator.next();
+		if (!next.done) {
+			frontier.push({ entry: next.value, iterator });
```
Comment (mtojek) [inline src/telemetry/export/files.ts]: if something fails, frontier entries won't be cleaned? should we add any try/catch/finally?

File: src/telemetry/export/files.ts [hunk 2]
```diff
@@ -50,17 +62,44 @@ export async function listTelemetryFilesForRange(
...
+		}
+	}
+
+	while (frontier.length > 0) {
+		frontier.sort((a, b) => compareEventEntries(a.entry, b.entry));
```
Comment (mtojek) [inline src/telemetry/export/files.ts]: Hm.. do we need to sort every iteration?

File: test/unit/telemetry/export/files.test.ts [hunk 3]
```diff
@@ -61,36 +61,71 @@ describe("listTelemetryFilesForRange", () => {
...
+	it("preserves events when a session timestamp moves backward", async () => {
 		writeFiles({
 			"telemetry-2026-05-12-aaaaaaaa.jsonl":
 				serializeTelemetryEventLine(
-					makeEvent({ timestamp: "2026-05-11T23:59:59.999Z" }),
+					makeSessionEvent("session-a", 0, "2026-05-12T10:02:00.000Z"),
 				) +
 				serializeTelemetryEventLine(
-					makeEvent({ timestamp: "2026-05-12T00:00:00.000Z" }),
+					makeSessionEvent("session-a", 1, "2026-05-12T10:00:00.000Z"),
```
Comment (mtojek) [inline test/unit/telemetry/export/files.test.ts]: should we keep the original test case, 23:59:59 and 00:00:00?

File: src/telemetry/export/files.ts [hunk 4]
```diff
@@ -50,17 +62,44 @@ export async function listTelemetryFilesForRange(
...
+		const cursor = frontier.shift();
+		if (!cursor) {
+			return;
```
Comment (mtojek) [inline src/telemetry/export/files.ts]: will the execution ever get here (see: `while (frontier.length > 0) {`) ?

File: src/telemetry/export/files.ts [hunk 5]
```diff
@@ -50,17 +62,44 @@ export async function listTelemetryFilesForRange(
...
-export async function* streamTelemetryEvents(
+/** Merge per-session append streams by timestamp, buffering one event per session. */
+export async function* streamTelemetryEventsSorted(
 	filePaths: readonly string[],
 	range: TelemetryDateRange,
 ): AsyncIterable<TelemetryEvent> {
-	for (const filePath of filePaths) {
-		const name = path.basename(filePath);
-		const stream = createReadStream(filePath, { encoding: "utf8" });
+	const frontier: EventCursor[] = [];
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts:72]: Fixed. Added try/finally cleanup so all opened session iterators are returned on abort, parse error, or early generator close.

File: src/telemetry/export/files.ts [hunk 6]
```diff
@@ -50,17 +62,44 @@ export async function listTelemetryFilesForRange(
 		.map(({ path: filePath }) => filePath);
 }
 
-/**
- * Yields events from `filePaths` in order, keeping only those whose timestamp
- * falls inside `range`. Reads line-by-line so memory stays flat on big files.
- */
-export async function* streamTelemetryEvents(
+/** Merge per-session append streams by timestamp, buffering one event per session. */
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. Updated the comment to state the monotonic-session assumption and append-order fallback for backward timestamps.

File: src/telemetry/export/files.ts [hunk 7]
```diff
@@ -96,15 +135,29 @@ export async function* streamTelemetryEvents(
...
 }
 
-function parseLogFilename(
-	dir: string,
-	name: string,
-): TelemetryLogFile | undefined {
-	const parsed = localJsonlFiles.parseFileName(name);
-	if (!parsed) {
-		return undefined;
+function sessionFileGroups(filePaths: readonly string[]): TelemetryLogFile[][] {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. Replaced the manual grouping loop with `Map.groupBy`

File: src/telemetry/export/files.ts [hunk 8]
```diff
@@ -25,7 +26,18 @@ interface TelemetryLogFile {
 	readonly part: number;
 }
 
-/** Log files that could contain events in `range`, in chronological order. */
+interface TelemetryEventEntry {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. Removed TelemetryEventEntry; the cursor now carries the event plus cached timestamp metadata

File: src/telemetry/export/files.ts [hunk 9]
```diff
@@ -114,3 +167,16 @@ function compareLogFiles(a: TelemetryLogFile, b: TelemetryLogFile): number {
 		a.part - b.part
 	);
 }
+
+function compareEventEntries(
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. Cached `timestampMs` when queueing events so comparisons do not reparse timestamps.

File: src/telemetry/export/files.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. Renamed the helper to `groupFilesBySession`

File: src/telemetry/export/files.ts [hunk 10]
```diff
@@ -114,3 +167,16 @@ function compareLogFiles(a: TelemetryLogFile, b: TelemetryLogFile): number {
...
+	a: TelemetryEventEntry,
+	b: TelemetryEventEntry,
+): number {
+	const timestamp =
```
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Removed

File: src/telemetry/export/files.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed. The stream comment now calls out range filtering

File: test/unit/telemetry/export/files.test.ts [hunk 11]
```diff
@@ -61,36 +61,71 @@ describe("listTelemetryFilesForRange", () => {
...
+			]),
+		).toEqual([
+			["2026-05-12T10:00:00.000Z", "session-a", 10],
+			["2026-05-12T10:00:00.000Z", "session-c", 0],
+			["2026-05-12T10:01:00.000Z", "session-b", 0],
+			["2026-05-12T10:02:00.000Z", "session-a", 11],
+		]);
+	});
+
+	it("preserves events when a session timestamp moves backward", async () => {
```
[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/export/files.test.ts]: Fixed. Restored range filtering coverage with the `23:59:59.999Z` boundary case.

[the remark this answers is not in the document]
Comment (EhabY) [reply test/unit/telemetry/export/files.test.ts]: Fixed. Added compound coverage where one session moves backward while another session interleaves

File: src/telemetry/export/files.ts (hunk 1 above)
Comment (EhabY) [reply src/telemetry/export/files.ts]: Fixed (reviewer flagged this as well)

File: src/telemetry/export/files.ts (hunk 2 above)
Comment (EhabY) [reply src/telemetry/export/files.ts]: We do not I suppose, replaced full frontier sorting with a single-pass minimum selection

File: test/unit/telemetry/export/files.test.ts (hunk 3 above)
Comment (EhabY) [reply test/unit/telemetry/export/files.test.ts]: Restored!

File: src/telemetry/export/files.ts (hunk 4 above)
Comment (EhabY) [reply src/telemetry/export/files.ts]: Removed the unreachable shift() guard by using `takeNextCursor`

**Review by mtojek — APPROVED** [review]
